File: c_project/components/web/views_sounds.c
#include "views.h"
#include "webserver.h"
#include "template_engine.h"
#include "persistent_dict.h"
#include "sound_manager.h"
#include "audio_player.h"
#include "lighting.h"
#include "json_helpers.h"
#include "cJSON.h"

#include <string.h>
#include <stdlib.h>

static int cmp_str(const void *a, const void *b)
{
    return strcmp(*(const char **)a, *(const char **)b);
}

/* Sound settings live on the current model, mirroring SoundManager.get_sounds()
 * in lib/sounds.py (models-based storage; legacy top-level "sounds" fallback
 * is not modeled in the C port since models are the only supported layout). */
static cJSON *get_sounds_dict(bool create_if_missing)
{
    cJSON *model = lighting_get_settings();
    if (!model) return NULL;

    cJSON *sounds = cJSON_GetObjectItem(model, "sounds");
    if (!sounds && create_if_missing) {
        sounds = cJSON_CreateObject();
        cJSON_AddItemToObject(model, "sounds", sounds);
    }
    return sounds;
}

static int get_current_volume(void)
{
    persistent_dict_t *store = persistent_dict_open(STORAGE_SYSTEM_SETTINGS_FILE);
    int volume = 20;
    if (store) {
        cJSON *v = persistent_dict_get_dup(store, "master_volume");
        if (v && cJSON_IsNumber(v)) volume = v->valueint;
        cJSON_Delete(v);
    }
    if (volume < 0) volume = 0;
    if (volume > 30) volume = 30;
    return volume;
}

/* Returns true if `title` appears in the sound_manager's currently-playing
 * titles array (mirrors _playing_sounds_by_title() membership check). */
static bool title_is_playing(cJSON *playing_arr, const char *title)
{
    if (!playing_arr || !title) return false;
    for (cJSON *item = playing_arr->child; item; item = item->next) {
        if (cJSON_IsString(item) && item->valuestring && strcmp(item->valuestring, title) == 0) return true;
    }
    return false;
}

/* Build a sorted "sounds" list of {title, file, high_quality, show_on_home,
 * is_playing, module_idx}. Mirrors _sounds_list()/_sounds_context(). When
 * include_playing is true, entries currently playing get is_playing=true
 * and module_idx set via sound_manager_get_playing_module(). */
static cJSON *build_sounds_list(cJSON *sounds_dict, bool include_playing, bool home_only)
{
    cJSON *result = cJSON_CreateArray();
    if (!sounds_dict) return result;

    int total = cJSON_GetArraySize(sounds_dict);
    if (total <= 0) return result;

    const char **names = calloc(total, sizeof(char *));
    if (!names) return result;

    int count = 0;
    for (cJSON *item = sounds_dict->child; item; item = item->next) {
        if (home_only && !json_get_bool(item, "show_on_home", true)) continue;
        names[count++] = item->string;
    }
    qsort(names, count, sizeof(char *), cmp_str);

    cJSON *playing = include_playing ? sound_manager_get_playing() : NULL;

    for (int i = 0; i < count; i++) {
        cJSON *sound = cJSON_GetObjectItem(sounds_dict, names[i]);
        cJSON *entry = cJSON_CreateObject();
        cJSON_AddStringToObject(entry, "title", names[i]);
        cJSON_AddNumberToObject(entry, "file", json_get_int(sound, "file", 0));
        cJSON_AddBoolToObject(entry, "high_quality", json_get_bool(sound, "high_quality", false));
        cJSON_AddBoolToObject(entry, "show_on_home", json_get_bool(sound, "show_on_home", true));
        bool is_playing = title_is_playing(playing, names[i]);
        cJSON_AddBoolToObject(entry, "is_playing", is_playing);
        cJSON_AddNumberToObject(entry, "module_idx",
                                is_playing ? sound_manager_get_playing_module(names[i]) : -1);
        cJSON_AddItemToArray(result, entry);
    }

    if (playing) cJSON_Delete(playing);
    free(names);
    return result;
}

/* Builds the {sounds, current_volume, sounds_poll_interval_seconds} context
 * shared by several views. Mirrors _sounds_context(). */
static void add_sounds_context(cJSON *ctx, bool include_playing, bool home_only)
{
    cJSON *sounds_dict = get_sounds_dict(false);
    cJSON *sounds_list = build_sounds_list(sounds_dict, include_playing, home_only);

    bool any_playing = false;
    if (include_playing) {
        for (cJSON *item = sounds_list->child; item; item = item->next) {
            if (cJSON_IsTrue(cJSON_GetObjectItem(item, "is_playing"))) { any_playing = true; break; }
        }
    }

    cJSON_AddItemToObject(ctx, "sounds", sounds_list);
    cJSON_AddNumberToObject(ctx, "current_volume", get_current_volume());
    cJSON_AddNumberToObject(ctx, "sounds_poll_interval_seconds", any_playing ? 5 : 30);
}

http_response_t *view_sounds(http_request_t *req)
{
    const char *action = request_get_form_field(req, "action");
    const char *sound_title = request_get_form_field(req, "sound_title");

    if (action && sound_title && sound_title[0]) {
        cJSON *sounds = get_sounds_dict(true);
        persistent_dict_t *store = persistent_dict_open(STORAGE_LIGHTING_SETTINGS_FILE);

        if (strcmp(action, "create_sound") == 0) {
            if (!cJSON_GetObjectItem(sounds, sound_title)) {
                cJSON *sound = cJSON_CreateObject();
                cJSON_AddNumberToObject(sound, "file", 1);
                cJSON_AddBoolToObject(sound, "high_quality", false);
                cJSON_AddBoolToObject(sound, "show_on_home", true);
                cJSON_AddNumberToObject(sound, "loop_count", 0);
                cJSON_AddNullToObject(sound, "chain_next");
                cJSON_AddItemToObject(sounds, sound_title, sound);
                persistent_dict_save(store);
            }
        } else if (strcmp(action, "delete_sound") == 0) {
            if (cJSON_GetObjectItem(sounds, sound_title)) {
                cJSON_DeleteItemFromObject(sounds, sound_title);
                persistent_dict_save(store);
            }
        }
    }

    cJSON *ctx = build_global_context();
    cJSON_AddStringToObject(ctx, "page_title", "Sounds");
    add_sounds_context(ctx, false, false);
    return webserver_render_response("setup/sounds.html", ctx);
}

http_response_t *view_sounds_summary(http_request_t *req)
{
    cJSON *ctx = build_global_context();
    add_sounds_context(ctx, false, false);
    return webserver_render_response("setup/sounds_summary.html", ctx);
}

http_response_t *view_sound_edit(http_request_t *req)
{
    cJSON *sounds = get_sounds_dict(false);

    if (req->method == HTTP_METHOD_GET) {
        const char *sound_title = request_get_query_param(req, "sound");
        cJSON *sound = (sound_title && sounds) ? cJSON_GetObjectItem(sounds, sound_title) : NULL;
        if (!sound_title || !sound) {
            return response_create(200, "text/html", "<p class=\"text-danger small\">Sound not found.</p>");
        }

        cJSON *ctx = build_global_context();
        cJSON_AddStringToObject(ctx, "sound_title", sound_title);
        cJSON_AddNumberToObject(ctx, "sound_file", json_get_int(sound, "file", 1));
        cJSON_AddBoolToObject(ctx, "sound_high_quality", json_get_bool(sound, "high_quality", false));
        cJSON_AddBoolToObject(ctx, "sound_show_on_home", json_get_bool(sound, "show_on_home", true));
        cJSON_AddNumberToObject(ctx, "sound_loop_count", json_get_int(sound, "loop_count", 0));
        const char *chain_next = json_get_string(sound, "chain_next", NULL);
        if (chain_next) cJSON_AddStringToObject(ctx, "sound_chain_next", chain_next);
        else cJSON_AddNullToObject(ctx, "sound_chain_next");

        int total = sounds ? cJSON_GetArraySize(sounds) : 0;
        const char **names = total > 0 ? calloc(total, sizeof(char *)) : NULL;
        int count = 0;
        if (names) {
            for (cJSON *item = sounds->child; item; item = item->next) names[count++] = item->string;
            qsort(names, count, sizeof(char *), cmp_str);
        }
        cJSON *available = cJSON_CreateArray();
        for (int i = 0; i < count; i++) cJSON_AddItemToArray(available, cJSON_CreateString(names[i]));
        free(names);
        cJSON_AddItemToObject(ctx, "available_sounds", available);

        return webserver_render_response("setup/sound_edit.html", ctx);
    }

    /* POST: update_sound */
    const char *action = request_get_form_field(req, "action");
    const char *old_sound_title = request_get_form_field(req, "old_sound_title");
    const char *sound_title = request_get_form_field(req, "sound_title");

    if (action && strcmp(action, "update_sound") == 0 && old_sound_title && old_sound_title[0] &&
        sound_title && sound_title[0] && sounds) {
        cJSON *sound = cJSON_GetObjectItem(sounds, old_sound_title);
        if (sound) {
            const char *file = request_get_form_field(req, "sound_file");
            const char *high_quality = request_get_form_field(req, "sound_high_quality");
            const char *show_on_home = request_get_form_field(req, "sound_show_on_home");
            const char *loop_count = request_get_form_field(req, "sound_loop_count");
            const char *chain_next = request_get_form_field(req, "sound_chain_next");

            cJSON_DeleteItemFromObject(sound, "file");
            cJSON_AddNumberToObject(sound, "file", file ? atoi(file) : 1);

            cJSON_DeleteItemFromObject(sound, "high_quality");
            cJSON_AddBoolToObject(sound, "high_quality", high_quality && strcmp(high_quality, "1") == 0);

            cJSON_DeleteItemFromObject(sound, "show_on_home");
            cJSON_AddBoolToObject(sound, "show_on_home", show_on_home && strcmp(show_on_home, "1") == 0);

            cJSON_DeleteItemFromObject(sound, "loop_count");
            cJSON_AddNumberToObject(sound, "loop_count", loop_count ? atoi(loop_count) : 0);

            cJSON_DeleteItemFromObject(sound, "chain_next");
            if (chain_next && chain_next[0]) cJSON_AddStringToObject(sound, "chain_next", chain_next);
            else cJSON_AddNullToObject(sound, "chain_next");

            if (strcmp(old_sound_title, sound_title) != 0) {
                if (!cJSON_GetObjectItem(sounds, sound_title)) {
                    cJSON *detached = cJSON_DetachItemFromObject(sounds, old_sound_title);
                    cJSON_AddItemToObject(sounds, sound_title, detached);
                }
                /* else: new title collides with an existing sound - keep old title,
                 * matching SoundEditView.post's silent no-rename fallback. */
            }

            persistent_dict_t *store = persistent_dict_open(STORAGE_LIGHTING_SETTINGS_FILE);
            persistent_dict_save(store);
        }
    }

    cJSON *ctx = build_global_context();
    cJSON_AddStringToObject(ctx, "page_title", "Sounds");
    add_sounds_context(ctx, false, false);
    return webserver_render_response("setup/sounds.html", ctx);
}

http_response_t *view_sounds_status(http_request_t *req)
{
    cJSON *ctx = build_global_context();
    add_sounds_context(ctx, true, true);
    return webserver_render_response("sounds/buttons.html", ctx);
}

http_response_t *view_play_sound(http_request_t *req)
{
    const char *title = request_get_form_field(req, "title");
    if (!title || !title[0]) {
        return response_create(200, "text/html",
            "<div class=\"alert alert-danger small py-2 mb-0\">No sound title provided</div>");
    }

    cJSON *sound = get_sounds_dict(false);
    cJSON *sound_def = sound ? cJSON_GetObjectItem(sound, title) : NULL;
    int file_number = sound_def ? json_get_int(sound_def, "file", 0) : 0;

    esp_err_t err = sound_manager_play(title, 0);
    if (err == ESP_ERR_NOT_FOUND) {
        char buf[192];
        snprintf(buf, sizeof(buf), "<div class=\"alert alert-danger small py-2 mb-0\">Sound '%s' not found</div>", title);
        return response_create(200, "text/html", buf);
    }
    if (err != ESP_OK) {
        return response_create(200, "text/html",
            "<div class=\"alert alert-danger small py-2 mb-0\">All audio modules are busy</div>");
    }

    int module_idx = sound_manager_get_playing_module(title);
    char buf[512];
    snprintf(buf, sizeof(buf),
        "<form hx-post=\"/sounds/stop\" hx-target=\"#sound-%d-result\" hx-swap=\"innerHTML\">"
        "<input type=\"hidden\" name=\"module_idx\" value=\"%d\">"
        "<input type=\"hidden\" name=\"title\" value=\"%s\">"
        "<input type=\"hidden\" name=\"file\" value=\"%d\">"
        "<button type=\"submit\" class=\"btn btn-sm btn-danger theme-sound-stop-btn\">&#9632; %s</button>"
        "</form>", file_number, module_idx, title, file_number, title);
    return response_create(200, "text/html", buf);
}

http_response_t *view_stop_sound(http_request_t *req)
{
    const char *title = request_get_form_field(req, "title");
    const char *file_str = request_get_form_field(req, "file");

    if (title && title[0]) {
        /* UI Stop uses the hardware-verified stop, matching StopSoundView ->
         * SoundManager.stop_sound() in Python; internal stops-lists keep the
         * fast sound_manager_stop(). */
        sound_manager_stop_verified(title);
    }

    int file_number = file_str ? atoi(file_str) : 0;
    char buf[384];
    snprintf(buf, sizeof(buf),
        "<form hx-post=\"/sounds/play\" hx-target=\"#sound-%d-result\" hx-swap=\"innerHTML\">"
        "<input type=\"hidden\" name=\"title\" value=\"%s\">"
        "<button type=\"submit\" class=\"btn btn-sm btn-outline-primary theme-sound-btn\">%s</button>"
        "</form>", file_number, title ? title : "", title ? title : "");
    return response_create(200, "text/html", buf);
}

http_response_t *view_stop_all_sounds(http_request_t *req)
{
    /* Mirrors StopAllSoundsView: stops audio player modules directly rather
     * than going through SoundManager (which also clears soundscape state) -
     * this is a deliberate asymmetry versus /soundscapes/stop in the Python
     * source. */
    audio_player_stop_all();

    cJSON *ctx = build_global_context();
    add_sounds_context(ctx, true, true);
    return webserver_render_response("sounds/buttons.html", ctx);
}

http_response_t *view_audio_volume(http_request_t *req)
{
    const char *volume_str = request_get_form_field(req, "master-volume");
    int volume = 20;
    if (volume_str) {
        volume = atoi(volume_str);
        if (volume < 0) volume = 0;
        if (volume > 30) volume = 30;
    }

    persistent_dict_t *store = persistent_dict_open(STORAGE_SYSTEM_SETTINGS_FILE);
    if (store) {
        persistent_dict_set(store, "master_volume", cJSON_CreateNumber(volume));
        persistent_dict_save(store);
    }

    audio_player_set_volume(volume);

    char buf[16];
    snprintf(buf, sizeof(buf), "%d/30", volume);
    return response_create(200, "text/plain", buf);
}

File: web/views_sounds.py
"""Views for managing sounds: list/edit sounds, playback control, status, and master volume."""

from webserver import View, render_template
from storage import PersistentDict

from web.views_common import (
    _sounds_context,
    _audio_debug_enabled,
    _get_model_scoped_dict,
    _save_model_scoped_dict,
)


def _get_sounds_dict(storage: PersistentDict) -> dict:
    """Get the sounds dictionary from storage."""

    scoped = _get_model_scoped_dict(storage, "sounds")
    if scoped is not None:
        return scoped

    if "sounds" not in storage:
        storage["sounds"] = {}
    return storage["sounds"]


def _save_sounds_dict(storage: PersistentDict, sounds: dict) -> None:
    """Save the sounds dictionary to storage."""

    if _save_model_scoped_dict(storage, "sounds", sounds):
        return

    storage["sounds"] = sounds
    storage.store()


class AudioVolumeView(View):
    """Handle volume control for audio players."""

    def post(self) -> str:
        """Set the master volume for all audio players (0-30).

        Reads the volume value from the form data (typically from a range input).
        Sends the volume command to all configured audio players and persists
        the setting for future power-ups.
        """

        try:
            volume_str = self.request.form_data.get("master-volume", "20")
            volume = int(volume_str)
            volume = max(0, min(30, volume))  # Clamp to valid range
        except (ValueError, TypeError):
            volume = 20

        # Persist the volume setting to disk
        storage: PersistentDict = PersistentDict()
        system_settings: dict = storage.get("system_settings", {})
        system_settings["master_volume"] = volume
        storage["system_settings"] = system_settings
        storage.store()  # Explicitly persist to disk

        # Send volume command to all active players and update the cached master volume
        try:
            from audio import AudioPlayer

            player = AudioPlayer()
            player.master_volume = volume
            for p in player.players:
                p.set_volume(volume)
        except Exception as e:
            print(f"AudioVolumeView: failed to set volume: {e}")

        return f"{volume}/30"


class SoundsSummaryView(View):
    """Return a summary snippet of sounds for the setup card."""

    def get(self) -> str:
        """Return sounds summary HTML fragment."""

        return render_template("setup/sounds_summary.html", _sounds_context())


class SoundsView(View):
    """List sounds and create or delete sounds."""

    def get(self) -> str:
        """Show sound list and create-sound form."""

        context: dict = _sounds_context()
        context["page_title"] = "Sounds"
        return render_template("setup/sounds.html", context)

    def post(self) -> str:
        """Create or delete a sound."""

        action: str = self.request.form_data.get("action", "").strip()
        sound_title: str = self.request.form_data.get("sound_title", "").strip()

        storage: PersistentDict = PersistentDict()
        sounds: dict = _get_sounds_dict(storage)

        if action == "create_sound" and sound_title:
            if sound_title not in sounds:
                sounds[sound_title] = {
                    "file": 1,
                    "high_quality": False,
                    "show_on_home": True,
                    "loop_count": 0,
                    "chain_next": None,
                }
                _save_sounds_dict(storage, sounds)

        elif action == "delete_sound" and sound_title and sound_title in sounds:
            del sounds[sound_title]
            _save_sounds_dict(storage, sounds)

        context: dict = _sounds_context()
        context["page_title"] = "Sounds"
        return render_template("setup/sounds.html", context)


class SoundEditView(View):
    """Edit a specific sound's properties."""

    def get(self) -> str:
        """Show edit form for a specific sound."""

        sound_title: str = self.request.query_params.get("sound", "").strip()

        storage: PersistentDict = PersistentDict()
        sounds: dict = _get_sounds_dict(storage)

        if not sound_title or sound_title not in sounds:
            return '<p class="text-danger small">Sound not found.</p>'

        sound: dict = sounds[sound_title]
        context: dict = {
            "sound_title": sound_title,
            "sound_file": sound.get("file", 1),
            "sound_high_quality": bool(sound.get("high_quality", False)),
            "sound_show_on_home": bool(sound.get("show_on_home", True)),
            "sound_loop_count": sound.get("loop_count", 0),
            "sound_chain_next": sound.get("chain_next"),
            "available_sounds": sorted(sounds.keys()),
        }
        return render_template("setup/sound_edit.html", context)

    def post(self) -> str:
        """Update a sound's properties."""

        action: str = self.request.form_data.get("action", "").strip()
        old_sound_title: str = self.request.form_data.get("old_sound_title", "").strip()
        sound_title: str = self.request.form_data.get("sound_title", "").strip()

        if action == "update_sound" and old_sound_title and sound_title:
            storage: PersistentDict = PersistentDict()
            sounds: dict = _get_sounds_dict(storage)

            if old_sound_title in sounds:
                # Get the current sound data
                sound_data: dict = sounds[old_sound_title]

                # Update with new form values
                try:
                    sound_data["file"] = int(self.request.form_data.get("sound_file", 1))
                except (ValueError, TypeError):
                    sound_data["file"] = 1

                sound_data["high_quality"] = self.request.form_data.get("sound_high_quality", "") == "1"
                sound_data["show_on_home"] = self.request.form_data.get("sound_show_on_home", "") == "1"

                try:
                    sound_data["loop_count"] = int(self.request.form_data.get("sound_loop_count", 0))
                except (ValueError, TypeError):
                    sound_data["loop_count"] = 0

                chain_next: str = self.request.form_data.get("sound_chain_next", "").strip()
                sound_data["chain_next"] = chain_next if chain_next else None

                # If title changed, delete old entry and create new one
                if old_sound_title != sound_title:
                    if sound_title not in sounds:
                        del sounds[old_sound_title]
                        sounds[sound_title] = sound_data
                    else:
                        # New title already exists, just update the old sound
                        sounds[old_sound_title] = sound_data
                else:
                    sounds[sound_title] = sound_data

                _save_sounds_dict(storage, sounds)

        context: dict = _sounds_context()
        context["page_title"] = "Sounds"
        return render_template("setup/sounds.html", context)


class PlaySoundView(View):
    """Handle requests to play a sound by title."""

    def post(self) -> str:
        """Play a sound and return a stop button on success.

        Form data:
            title: Sound title to play
        """

        title: str = self.request.form_data.get("title", "").strip()

        if not title:
            return '<div class="alert alert-danger small py-2 mb-0">No sound title provided</div>'

        try:
            from sounds import SoundManager

            manager: SoundManager = SoundManager()
            sound_info = None
            try:
                sound_info = manager.get_sound_by_title(title)
            except Exception:
                pass

            module_idx: int = manager.play_sound(title)
            file_number: int = sound_info.get("file", 0) if isinstance(sound_info, dict) else 0
            return (
                f'<form hx-post="/sounds/stop"'
                f' hx-target="#sound-{file_number}-result"'
                f' hx-swap="innerHTML">'
                f'<input type="hidden" name="module_idx" value="{module_idx}">'
                f'<input type="hidden" name="title" value="{title}">'
                f'<input type="hidden" name="file" value="{file_number}">'
                f'<button type="submit" class="btn btn-sm btn-danger theme-sound-stop-btn">'
                f"&#9632; {title}"
                f"</button>"
                f"</form>"
            )
        except ImportError:
            return '<div class="alert alert-warning small py-2 mb-0">Audio system not available</div>'
        except ValueError:
            return '<div class="alert alert-danger small py-2 mb-0">' f"Sound '{title}' not found" "</div>"
        except Exception as err:
            from audio import NoPlayersAvailable

            if isinstance(err, NoPlayersAvailable):
                return '<div class="alert alert-danger small py-2 mb-0">' "All audio modules are busy" "</div>"
            else:
                return '<div class="alert alert-danger small py-2 mb-0">' f"Error: {str(err)}" "</div>"


class StopSoundView(View):
    """Handle requests to stop a specific audio module."""

    def post(self) -> str:
        """Stop the given module and restore the play button.

        Form data:
            module_idx: Index of the audio module to stop
            title: Sound title (to restore the play button label)
            file: File number (used to restore the correct hx-target)
        """

        module_idx_str: str = self.request.form_data.get("module_idx", "")
        title: str = self.request.form_data.get("title", "").strip()
        file_str: str = self.request.form_data.get("file", "0")

        try:
            from sounds import SoundManager

            manager: SoundManager = SoundManager()
            module_idx = None
            try:
                parsed_module_idx: int = int(module_idx_str)
                module_idx = parsed_module_idx
            except (ValueError, TypeError):
                module_idx = None

            stop_ok: bool = manager.stop_sound(title, module_idx)
            if not stop_ok:
                print(f"StopSoundView: stop command did not confirm stopped for title='{title}' module={module_idx}")
        except Exception as err:
            print(f"StopSoundView: error stopping module: {err}")

        try:
            file_number: int = int(file_str)
        except (ValueError, TypeError):
            file_number = 0

        return (
            f'<form hx-post="/sounds/play"'
            f' hx-target="#sound-{file_number}-result"'
            f' hx-swap="innerHTML">'
            f'<input type="hidden" name="title" value="{title}">'
            f'<button type="submit" class="btn btn-sm btn-outline-primary theme-sound-btn">'
            f"{title}"
            f"</button>"
            f"</form>"
        )


class StopAllSoundsView(View):
    """Handle requests to stop all audio modules."""

    def post(self) -> str:
        """Stop all audio modules and return the re-rendered sounds buttons section."""

        try:
            from audio import AudioPlayer

            audio_player: AudioPlayer = AudioPlayer()
            audio_player.stop_all()
        except Exception as err:
            print(f"StopAllSoundsView: error stopping all: {err}")

        return render_template("sounds/buttons.html", _sounds_context(include_playing=True, home_only=True))


class SoundsStatusView(View):
    """Return the home sounds controls fragment with current playback state."""

    def get(self) -> str:
        """Render sound buttons using current playing/not-playing state."""

        debug_enabled: bool = _audio_debug_enabled()
        print(f"sounds-status: endpoint hit debug_enabled={debug_enabled}")
        if debug_enabled:
            print("audio-debug: sounds_status endpoint hit")

        context: dict = _sounds_context(include_playing=True, home_only=True)
        if debug_enabled:
            try:
                sound_count: int = len(context.get("sounds", []))
            except Exception:
                sound_count = -1
            print(f"audio-debug: sounds_status render sound_count={sound_count}")

        return render_template("sounds/buttons.html", context)
